rust/lance/src/io/exec/scalar_index.rs

```diff
@@ -0,0 +1,325 @@
+        Partitioning::RoundRobinBatch(1)
+    }
+
+    fn output_ordering(&self) -> Option<&[datafusion::physical_expr::PhysicalSortExpr]> {
+        // No guarantee a scalar index scan will return row ids in any meaningful order
+        None
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        _children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> datafusion::error::Result<Arc<dyn ExecutionPlan>> {
+        todo!()
+    }
+
+    fn execute(
+        &self,
+        _partition: usize,
+        _context: Arc<datafusion::execution::context::TaskContext>,
+    ) -> datafusion::error::Result<datafusion::physical_plan::SendableRecordBatchStream> {
+        let batch_fut = Self::do_execute(self.expr.clone(), self.dataset.clone());
+        let stream = futures::stream::iter(vec![batch_fut])
+            .then(|batch_fut| batch_fut.map_err(|err| err.into()))
+            .boxed()
+            as BoxStream<'static, datafusion::common::Result<RecordBatch>>;
+        Ok(Box::pin(StreamWithSchema {
+            schema: SCALAR_INDEX_SCHEMA.clone(),
+            stream,
+        }))
+    }
+
+    fn statistics(&self) -> datafusion::physical_plan::Statistics {
+        todo!()
+    }
+}
+
+lazy_static::lazy_static! {
+    pub static ref MATERIALIZE_INDEX_SCHEMA: SchemaRef = Arc::new(Schema::new(vec![ROW_ID_FIELD.clone()]));
+}
+
+/// An execution node that performs a scalar index search and materializes the mask into row ids
+///
+/// First, the index is searched to determine the mask that should be applied.  Then, we take the
+/// list of fragments, iterate through all possible row ids, and materialize the row ids that satisfy
+/// the mask.  The output of this node is a list of row ids suitable for use in a take operation.
+#[derive(Debug)]
+pub struct MaterializeIndexExec {
+    dataset: Arc<Dataset>,
+    expr: ScalarIndexExpr,
+    fragments: Arc<Vec<Fragment>>,
+}
+
+impl DisplayAs for MaterializeIndexExec {
+    fn fmt_as(&self, t: DisplayFormatType, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match t {
+            DisplayFormatType::Default | DisplayFormatType::Verbose => {
+                write!(f, "MaterializeIndex: query={}", self.expr)
+            }
+        }
+    }
+}
+
+struct FragIdIter {
+    src: Arc<Vec<Fragment>>,
+    frag_idx: usize,
+    idx_in_frag: usize,
+}
+
+impl FragIdIter {
+    fn new(src: Arc<Vec<Fragment>>) -> Self {
+        Self {
+            src,
+            frag_idx: 0,
+            idx_in_frag: 0,
+        }
+    }
+}
+
+impl Iterator for FragIdIter {
+    type Item = u64;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        while self.frag_idx < self.src.len() {
+            let frag = &self.src[self.frag_idx];
+            if self.idx_in_frag
+                < frag
+                    .physical_rows
+                    .expect("Fragment doesn't have physical rows recorded")
```

Comment:
Are we retrieving this earlier?

Reply:
Are you asking if there is a safeguard that keeps us from getting to this point if one of the fragments is missing physical rows?  If so, then no, but it is a good idea and I will add it.

Reply:
In `create_plan` we now do an early check.  If any of the fragments are missing the physical row counts we will fallback to the non-indexed approach.